spark-common/src/main/js/src_contentwin/spark-contentwindow.js

```diff
@@ -3,6 +3,90 @@ import 'iframe-resizer/js/iframeResizer.contentWindow';
+            // copy theming CSS from outer window
+            parentWindow.document.querySelectorAll('style[data-theme]:not([data-custom-theme])')
+                .forEach((themeStyle) => {
+                    var theme = themeStyle.getAttribute('data-theme');
+
+                    var iframeThemeStyle = document.createElement('style');
+                    iframeThemeStyle.setAttribute('data-theme', theme);
+                    iframeThemeStyle.textContent = themeStyle.textContent;
+
+                    document.head.appendChild(iframeThemeStyle);
+                })
+        }
+    }
+
+    /*
+     * Copy the data attributes from the parent pages HTML element to this one.
+     * Setting those effectively enables theming (as long as the required token styles are available in this frame).
+     * See https://developer.atlassian.com/server/framework/atlassian-theme/html/ for the meaning of those data attributes.
+     */
+    function copyThemeAttributes() {
+        if (parentWindow && parentWindow.document) {
+            var parentHtml = parentWindow.document.documentElement;
+
+            if (parentHtml) {
+                var parentData = parentHtml.dataset;
+                setThemeAttributes(parentData.colorMode, parentData.theme);
+            }
+        }
+    }
+
+    function setThemeAttributes(colorMode, themeKeys) {
+        var thisHtml = document.documentElement;
+        thisHtml.dataset.colorMode = colorMode;
+        thisHtml.dataset.theme = themeKeys;
+    }
+
+    function handleThemeChange(event) {
+        try {
+            const { type, colorMode, darkThemeKey, lightThemeKey } = JSON.parse(event.data);
+            if (type === "theme.change") {
+                // parent pages theme attributes are not yet updated, construct attributes from event payload instead
+                var themeKeys = 'dark:' + darkThemeKey + ' light:' + lightThemeKey;
+                // colorMode is uppercase and in the case of 'AUTO' needs to be resolved manually
+                var colorModeResolved = (colorMode === 'AUTO') ? getPreferredTheme() : colorMode.toLowerCase();
+
+                setThemeAttributes(colorModeResolved, themeKeys);
+            }
+        } catch (e) { /* ignore, usually due to event.data not being a parseable JSON string */ }
+    }
+
+    /*
+     * Returns the users preferred color schema, falls back to light if that is not possible
+     */
+    function getPreferredTheme() {
+        var isMatchMediaAvailable = 'matchMedia' in windowEl;
+        if (isMatchMediaAvailable) {
+            var darkModeMediaQuery = '(prefers-color-scheme: dark)';
+            var darkModeMql = isMatchMediaAvailable && windowEl.matchMedia(darkModeMediaQuery);
+            return darkModeMql.matches ? 'dark' : 'light';
+        }
+        return 'light';
+    }
+
+    function initializeTheming() {
+        copyThemeTokens();
+        copyThemeAttributes();
+
+        windowEl.addEventListener("message", handleThemeChange);
```

Comment:
who's sending this message? 🤔

Reply:
That's send by [Confluence](https://community.developer.atlassian.com/t/dark-theme-eap-in-confluence-data-center/76551#iframe-tags-5) when using the theme switcher. I'll add a comment there for clarity!